text/0033-crash-recovery.md

```diff
@@ -0,0 +1,108 @@
+
+However, if a network is stalled, the new era won't be able to begin, because there aren't enough validators to finish the current one. Thus, even if enough validators come back online, this currently doesn't help: they won't vote until a new era begins, and this will never happen.
+
+This CEP aims to propose a solution that would allow the validators to start voting in eras that are already in progress, making it possible for the network to resume normal operation after temporary issues.
+
+**Note 2020-01-11:** We currently have a quick workaround in place, which basically does what is described below, except it doesn't cater to endorsements, nor does it use storage - it bypasses the storage component by directly creating additional files. While working in practice for the moment, it should be replaced by a proper solution.
+
+## Guide-level explanation
+
+[guide-level-explanation]: #guide-level-explanation
+
+Since the main problem is the risk of creating an accidental equivocation, we need to make sure that a rejoining validator will be aware of all its previous messages in the relevant era.
+
+For the units, it is enough to save the hash of the latest created unit. This would make it possible to download the unit from other validators along with its whole history and be able to verify the correctness of downloaded data.
+
+Units aren't the only thing that might be used to create an equivocation in Highway, though. Endorsements are such a thing, too. It would therefore be necessary to save the hashes of all endorsed units, too. When rejoining the era, the validator would read the list of such hashes, download the relevant units and make sure that it never endorses conflicting units in the future.
+
+The only remaining detail is how to implement saving the relevant hashes in practice. The simplest solution of just saving the hash whenever we create a unit or endorsement has a significant drawback. Imagine we create a unit, save its hash as the hash of the last unit created by us, and then crash before sending the unit to any validator (for any reason, eg. a power outage). In such a case, after restarting, we wouldn't be able to download it again. We wouldn't be able to make sure that we don't equivocate and so we wouldn't start voting again. It is probably the most secure solution, though. An alternative will be discussed in [Rationale and alternatives](#rationale-and-alternatives).
+
+## Reference-level explanation
+
+[reference-level-explanation]: #reference-level-explanation
+
+The consensus component should:
+
+- Whenever a unit or an endorsement is created, emit a request to storage to store the latest data for the era. The data stored would contain:
+    - The hash of the latest unit
+    - The list of hashes of endorsed units
```

Comment:
It would even suffice to, for each validator, store only the hash of their _latest_ unit that we endorsed.